### PR TITLE
refactor(engine-core): Centralizing diffing algo logic

### PR DESCRIPTION
**Pull request by pmdartus**
## Details

This PR centralizes all the diffing algorithm logic to `hooks.ts`. Prior to this change, the logic was split between `hooks.ts` and `api.ts`. The `api.ts` is now solely focused on VNodes creation.

Mics changes:
- Move `3rdparty/snabbdom/types` to `framework/vnodes.ts`
- Externalize elements and attribute namespaces to `@lwc/shared`

This PR is part of the larger refactor for [coupling the rehydration logic from the diffing logic](https://github.com/salesforce/lwc/pull/2608). This is the last setup PR prior to teasing apart the patching logic from the hydration logic. 

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

* ✅ No, it does not introduce an observable change.

## GUS work item
W-10409559

### CODE REVIEW

**Review by nolanlawson — APPROVED** [review]

File: packages/@lwc/engine-core/src/framework/hooks.ts [hunk 1]
```diff
@@ -461,3 +755,185 @@ export function markAsDynamicChildren(children: VNodes) {
 function hasDynamicChildren(children: VNodes): boolean {
     return FromIteration.has(children);
 }
+
+function createKeyToOldIdx(
```
Comment (caridy) [inline packages/@lwc/engine-core/src/framework/hooks.ts:759]: @pmdartus was there any change to this block other than copying it from snabbdom.ts?

File: packages/@lwc/engine-core/src/framework/hooks.ts [hunk 2]
```diff
@@ -34,6 +64,263 @@ import { applyEventListeners } from './modules/events';
 import { applyStaticClassAttribute } from './modules/static-class-attr';
 import { applyStaticStyleAttribute } from './modules/static-style-attr';
 
+const enum LwcDomMode {
```
Comment (caridy) [inline packages/@lwc/engine-core/src/framework/hooks.ts:67]: same q here, is there any functional change in this block other than just copying it from the api.ts file?

**Review by caridy — APPROVED** [review]
In principle, this looks good to me. Waiting for the answers to my two soft questions.

File: packages/@lwc/engine-core/src/framework/hooks.ts (hunk 1 above)
Comment (pmdartus) [reply packages/@lwc/engine-core/src/framework/hooks.ts:759]: This is a pure copy-paste. No change was made to this routine.

File: packages/@lwc/engine-core/src/framework/hooks.ts (hunk 2 above)
Comment (pmdartus) [reply packages/@lwc/engine-core/src/framework/hooks.ts:67]: Same observation. No change was made to this routine.

**Comment by pmdartus** [comment]
Nucleus is failing because of a Nexus hiccup.

**Pull request merged**